pkg/controllers/nodeclass/status/subnet.go

```diff
@@ -44,55 +47,77 @@ func NewSubnetReconciler(subnetClient instance.SubnetsAPI) *SubnetReconciler {
 
 	// TODO(bsoghigian): Support SubnetFull readiness reason
 	SubnetUnreadyReasonFull = "SubnetFull"
 
+	// TODO(bsoghigian): check if the cidr of a subnet-id is overlapping with any of the static agentpools,
+	// AKSNodeClass subnets, or any defaulting reserved networking addresses for AKS (--dns-service-ip)
+	SubnetUnreadyReasonCIDROverlapping = "SubnetCIDROverlapping"
+
+	// TODO(bsoghigian): check cluster identity has rbac for subnet/read subnet/join for a given vnetSubnetID
+	SubnetUnreadyReasonRBACInvalid = "SubnetRBACInvalid"
+)
+
+const (
+	// TODO: Use this in SubnetFull logic
 	// Azure reserves the first four addresses and the last address, for a total of 5 ips for each subnet.
 	// For example 192.168.1.0/24 has the following reserved addresses
 	// 192.168.1.0 Network Address
 	// 192.168.1.1 Reserved by azure for the default gateway
 	// 192.168.1.2, 192.168.1.3: Reserved by Azure to map the Azure DNS IP addresses to the virtual network space.
 	// 192.168.1.255: Network broadcast address
 	AzureReservedIPs = 5
-
-	// TODO(bsoghigian): check if the cidr of a subnet-id is overlapping with any of the static agentpools,
-	// AKSNodeClass subnets, or any defaulting reserved networking addresses for AKS (--dns-service-ip)
-	SubnetUnreadyReasonCIDROverlapping = "SubnetCIDROverlapping"
-
-	// TODO(bsoghigian): check cluster identity has rbac for subnet/read subnet/join for a given vnetSubnetID
-	SubnetUnreadyReasonRBACInvalid = "SubnetRBACInvalid"
 )
 
 func (r *SubnetReconciler) Reconcile(ctx context.Context, nodeClass *v1beta1.AKSNodeClass) (reconcile.Result, error) {
-	return r.validateVNETSubnetID(ctx, nodeClass)
 	// TODO: Handle podSubnetID readiness here as well
-	// Access to state from the cluster only needs to be retrieved once for cidr validation
+	result, err := r.validateVNETSubnetID(ctx, nodeClass)
+	return result, err
 }
 
 func (r *SubnetReconciler) validateVNETSubnetID(ctx context.Context, nodeClass *v1beta1.AKSNodeClass) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+	clusterSubnetID := options.FromContext(ctx).SubnetID
 	subnetID := lo.Ternary(nodeClass.Spec.VNETSubnetID != nil, lo.FromPtr(nodeClass.Spec.VNETSubnetID), options.FromContext(ctx).SubnetID)
-	subnetComponents, err := utils.GetVnetSubnetIDComponents(subnetID)
+	logger.V(2).Info("validating os.spec.vnetSubnetID: %s", subnetID)
+
+	nodeClassSubnetComponents, err := utils.GetVnetSubnetIDComponents(subnetID)
 	if err != nil {
 		nodeClass.StatusConditions().SetFalse(
 			v1beta1.ConditionTypeSubnetReady,
 			SubnetUnreadyReasonIDInvalid,
 			fmt.Sprintf("Failed to parse vnetSubnetID %s", subnetID),
 		)
-		return reconcile.Result{RequeueAfter: time.Minute}, err
+		return reconcile.Result{}, err
+	}
+	if subnetID != clusterSubnetID {
+		clusterSubnetIDParts, _ := utils.GetVnetSubnetIDComponents(clusterSubnetID) // Assume valid cluster subnet id
+		if !clusterSubnetIDParts.IsSameVNET(nodeClassSubnetComponents) {
```

Comment:
So, one question here. 

Under the existing system, without this validation, could a customer currently have a valid setup, that would be provisioning correctly, where the vnet doesn't match for `options.FromContext(ctx).SubnetID` and `nodeClass.Spec.VNETSubnetID`, but both are set?

Two separate vnets in mutual existence for different cases?

Main reason I'm asking is that, now we are in a GA state, want to be more careful on potential breakages, if a potentially valid use case customers might be using.

Reply:
Yes but its always been an unsupported configuration and documented as such.